tools/scripts/generate_rp_boot2.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python3
+# Copyright (c) 2018, Niklas Hauser
+#
+# This file is part of the modm project.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import os
+import fnmatch
+import shutil
+import subprocess
+import binascii
+import struct
+import sys
+import multiprocessing
+from pathlib import Path
+
+def repopath(path):
+	return (Path(os.path.abspath(__file__)).parents[2] / path)
+
+def run(where, command, stdin=None):
+	print(command)
+	result = subprocess.run(command, shell=True, cwd=where, input=stdin, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+	if result.returncode != 0:
+		print('failed exec command ' + command)
+		print(result.stderr.decode("utf-8").strip(" \n"))
+		exit(1)
+	return (result.returncode,
+	        result.stdout.decode("utf-8").strip(" \n"),
+	        result.stderr.decode("utf-8").strip(" \n"))
+
+def bitrev(x, width):
+    return int("{:0{w}b}".format(x, w=width)[::-1], 2)
+
+boot2_variants = [
+	"generic_03h",
+	"at25sf128a",
+	"is25lp080",
+	"w25q080",
+	"w25x10cl"
+]
+
+for variant in boot2_variants:
+	
+	builddir = repopath('build/boot2_' + variant)
+	if not os.path.exists(builddir):
+		os.mkdir(builddir)
+	run(builddir,"cmake ../../ext/rp/pico-sdk -G \"Unix Makefiles\" -DPICO_DEFAULT_BOOT_STAGE2=boot2_" + variant)
+	run(builddir,'make bs2_default_padded_checksummed_asm') # for validate
+	#run(builddir,'make bs2_default_bin')
+	ifile = os.path.join(builddir,'src/rp2_common/boot_stage2/bs2_default.bin')
+	try:
+	    idata = open(ifile, "rb").read()
+	except:
+		sys.exit("Could not open input file '{}'".format(ifile))
+	pad = 256
+	if len(idata) >= pad - 4:
+		sys.exit("Input file size ({} bytes) too large for final size ({} bytes)".format(len(idata), pad))
+	idata_padded = idata + bytes(pad - 4 - len(idata))
+	seed = 0xffffffff
+	# Our bootrom CRC32 is slightly bass-ackward but it's best to work around for now (FIXME)
+	# 100% worth it to save two Thumb instructions
+	checksum = bitrev(
+	    (binascii.crc32(bytes(bitrev(b, 8) for b in idata_padded), seed ^ 0xffffffff) ^ 0xffffffff) & 0xffffffff, 32)
+	odata = idata_padded + struct.pack("<L", checksum)
+
+	ofilename = repopath('src/modm/platform/core/rp/boot2_' + variant + '.cpp')
+	try:
+		with open(ofilename, "w") as ofile:
+			ofile.write("// Stage2 bootloadr\n\n")
+			ofile.write("#include <cstdint>\n")
+			ofile.write('#include <modm/architecture/utils.hpp>\n')
+			ofile.write("extern \"C\" modm_section(\".boot2\") const uint8_t boot2[256] = {\n")
+			for offs in range(0, len(odata), 16):
+				chunk = odata[offs:min(offs + 16, len(odata))]
+				ofile.write("\t {},\n".format(", ".join("0x{:02x}".format(b) for b in chunk)))
+			ofile.write("};\n")
+	except:
+		sys.exit("Could not open output file '{}'".format(ofilename))
+
+exit(0)
```

Comment:
This should be generated into the pico-sdk-partial so that it's automatically updated. Otherwise it's another step for me to forget when I update the pico-sdk submodule in ext 😬